account-kit/smart-contracts/src/ma-v2/account/account.ts

```diff
@@ -0,0 +1,158 @@
+import type {
+  EntryPointDef,
+  SmartAccountSigner,
+  SmartContractAccountWithSigner,
+  ToSmartContractAccountParams,
+} from "@aa-sdk/core";
+import {
+  createBundlerClient,
+  getAccountAddress,
+  getEntryPoint,
+  toSmartContractAccount,
+} from "@aa-sdk/core";
+import {
+  concatHex,
+  encodeFunctionData,
+  getContract,
+  type Address,
+  type Chain,
+  type Hex,
+  type Transport,
+} from "viem";
+import { accountFactoryAbi } from "../abis/accountFactoryAbi.js";
+import { addresses } from "../utils.js";
+import { standardExecutor } from "../../msca/account/standardExecutor.js";
+import { singleSignerMessageSigner } from "../modules/single-signer-validation/signer.js";
+
+export const DEFAULT_OWNER_ENTITY_ID = 0;
+
+export type SMAV2Account<
+  TSigner extends SmartAccountSigner = SmartAccountSigner
+> = SmartContractAccountWithSigner<"SMAV2Account", TSigner, "0.7.0">;
+
+export type CreateSMAV2AccountParams<
+  TTransport extends Transport = Transport,
+  TSigner extends SmartAccountSigner = SmartAccountSigner,
+  TEntryPointVersion extends "0.7.0" = "0.7.0"
+> = Pick<
+  ToSmartContractAccountParams<
+    "SMAV2Account",
+    TTransport,
+    Chain,
+    TEntryPointVersion
+  >,
+  "transport" | "chain"
+> & {
+  signer: TSigner;
+  salt?: bigint;
+  factoryAddress?: Address;
+  initCode?: Hex;
+  initialOwner?: Address;
+  accountAddress?: Address;
+  entryPoint?: EntryPointDef<TEntryPointVersion, Chain>;
+  isGlobalValidation?: boolean;
+  entityId?: bigint;
+};
+
+export async function createSMAV2Account<
+  TTransport extends Transport = Transport,
+  TSigner extends SmartAccountSigner = SmartAccountSigner
+>(
+  config: CreateSMAV2AccountParams<TTransport, TSigner>
+): Promise<SMAV2Account<TSigner>>;
+
+export async function createSMAV2Account(
+  config: CreateSMAV2AccountParams
+): Promise<SMAV2Account> {
+  const {
+    transport,
+    chain,
+    signer,
+    salt = 0n,
+    factoryAddress = addresses.accountFactory,
+    initCode,
+    initialOwner,
+    accountAddress,
+    entryPoint = getEntryPoint(chain, { version: "0.7.0" }),
+    isGlobalValidation = true,
+    entityId = 0n,
+  } = config;
+
+  if (entityId >= 2n ** 32n) {
+    throw new Error("Entity ID must be less than uint32.max");
+  }
+
+  const client = createBundlerClient({
+    transport,
+    chain,
+  });
+
+  const getAccountInitCode = async () => {
+    if (initCode) {
+      return initCode;
+    }
+
+    // If an initial owner is not provided, use the signer's address
+    const ownerAddress = initialOwner || (await signer.getAddress());
+
+    return concatHex([
+      factoryAddress,
+      encodeFunctionData({
+        abi: accountFactoryAbi,
+        functionName: "createAccount",
+        args: [ownerAddress, salt, DEFAULT_OWNER_ENTITY_ID],
```

Comment:
looking at this, this kinda answers one of my earlier questions: it sounds like we can have some stricter typing around passing in an `entityId` and `initCode` override. 

For example, if entityId > 0, then initCode is required as an override and then we don't have to pass in the initial `owner` here

Reply:
Thinking a little more - we default to entityId = 0 if no entityId is specified, but it's possible for someone to create an account with the owner associated to some entityId != 0 (not the best idea though). But because of that, we probably can't apply stricter typing around that

Reply:
whoops - just realized this the above is for MA and not SMA, can disregard